classes/local/store/object_file_system.php

```diff
@@ -1154,4 +1163,53 @@ private function update_object(array $result): array {
 
         return $result;
     }
+
+    /**
+     * Pushes tags to the external store (post upload) for a given hash.
+     * External client must support tagging.
+     *
+     * @param string $contenthash file to sync tags for
+     */
+    public function push_object_tags(string $contenthash) {
+        if (!$this->get_external_client()->supports_object_tagging()) {
+            throw new coding_exception("Cannot sync tags, external client does not support tagging.");
+        }
+
+        // Get a lock before syncing, to ensure other parts of objectfs are not moving/interacting with this object.
+        // Don't wait for it, we want to fail fast.
+        $lock = $this->acquire_object_lock($contenthash, 0);
+
+        // No lock - just skip it.
+        if (!$lock) {
+            throw new coding_exception("Could not get object lock");
+        }
+
+        try {
+            $objectexists = $this->is_file_readable_externally_by_hash($contenthash);
+
+            // Object must exist, and we can overwrite (and not care about existing tags)
+            // or cannot overwrite, and the tags are empty.
+            // Avoid unnecessarily checking tags, since this is an extra API call.
+            $canset = $objectexists && (tag_manager::can_overwrite_object_tags() ||
+                empty($this->get_external_client()->get_object_tags($contenthash)));
+
+            if ($canset) {
+                $tags = tag_manager::gather_object_tags_for_upload($contenthash);
+                $this->get_external_client()->set_object_tags($contenthash, $tags);
+                tag_manager::store_tags_locally($contenthash, $tags);
+                tag_manager::record_tag_pushed_time($contenthash, time());
+            }
+
+            // Regardless, it has synced.
+            tag_manager::mark_object_tag_sync_status($contenthash, tag_manager::SYNC_STATUS_COMPLETE);
```

Comment:
Correct, it was just simpler (code-wise) that way. But since you mentioned it I looked at it again and managed to make it do one update call with a tiny bit of extra code complexity. :+1: